File: ruby/linked-list/linked_list.rb
# The Deque class that we are being asked to implement could be done
# very easily using a Ruby array, so we will avoid that if possible

class Deque
  attr_reader :head, :tail

  def initialize
    @head = @tail = nil
  end

  def push(x)
    new_tail = Node.new(x)

    if @tail
      @tail.next = new_tail
      new_tail.prev = @tail
    end

    @tail = new_tail

    @head = @tail unless @head
  end

  def pop
    return unless @tail

    old_tail = @tail
    new_tail = @tail.prev

    if new_tail.nil?
      @head = @tail = nil
    else
      new_tail.next = nil
      @tail = new_tail
    end

    return old_tail.value
  end

  # shift is like pop, but it operates on the head instead of the tail
  def shift
    return unless @head

    old_head = @head
    new_head = @head.next

    if new_head.nil?
      @head = @tail = nil
    else
      new_head.prev = nil
      @head = new_head
    end

    return old_head.value
  end

  # unshift is like push, but it operates on the head instead of the
  # tail
  def unshift(x)
    new_head = Node.new(x)

    if @head
      @head.prev = new_head
      new_head.next = @head
    end

    @head = new_head

    @tail = @head unless @tail
  end
end

class Node
  attr_accessor :value, :next, :prev

  def initialize(value)
    @value = value
    @next = @prev = nil
  end
end

File: javascript/linked-list/linked-list.js
class ListNode {
    constructor(n) {
        this.value = n;
        this.next = this.prev = null;
    }

    join(other) {
        this.next = other;
        other.prev = this;
    }
}

export class LinkedList {
    constructor() {
        this.head = this.tail = null;
        this.size = 0;
    }

    push(n) {
        if (this.size) {
            this.tail.join(new ListNode(n));
            this.tail = this.tail.next;
        } else {
            this.head = this.tail = new ListNode(n);
        }

        this.size += 1;
    }

    pop() {
        if (this.size == 0) return;

        let output = this.tail.value;

        this.tail = this.tail.prev;

        if (this.tail)
            this.tail.next = null;

        this.size -= 1;

        return output;
    }

    shift() {
        if (this.size == 0) return;

        let output = this.head.value;

        this.head = this.head.next;

        if (this.head)
            this.head.prev = null;

        this.size -= 1;

        return output;
    }

    unshift(n) {
        if (this.size) {
            new ListNode(n).join(this.head);
            this.head = this.head.prev;
        } else {
            this.head = this.tail = new ListNode(n);
        }

        this.size += 1;
    }

    // Searches for the first node with the given value -- starting from
    // the head -- and returns it
    find(n) {
        let curr = this.head;

        while (curr && curr.value != n)
            curr = curr.next;

        return curr;
    }

    delete(n) {
        if (this.size == 0) return;

        let node = this.find(n);

        if (!node) return;

        if (node.prev)
            node.prev.next = node.next;

        if (node.next)
            node.next.prev = node.prev;

        if (node == this.head)
            this.head = node.next;

        if (node == this.tail)
            this.tail = node.prev;

        this.size -= 1;
    }

    count() {
        return this.size;
    }
}
